### PR TITLE
Constructor Comparison Update

### PR DESCRIPTION
**Pull request by eslutz**
## Description

Updated the `ToCompare()` method in `ConstructorOrderInfo` to also compare the parameter types during comparison.

## Type of change: 2

1. Non-functional change (e.g. documentation changes, removing unused `using` directives, renaming local variables, etc)
2. Bug fix (non-breaking change that fixes an issue)
3. New feature (non-breaking change that adds functionality)
4. Breaking change (fix or feature that could cause existing functionality to not work as expected)

---

<sup>_[Reviewer guidelines](../blob/main/CONTRIBUTING.md#reviewing-changes)_</sup>

### CODE REVIEW

**Comment by NeilSchiefer** [comment]
This may be the same thing as the current failing test.  But I was thinking we should have a unit test for this scenario - https://github.com/RockLib/RockLib.Configuration/blob/main/RockLib.Configuration.ObjectFactory/ConfigurationObjectFactory.cs#L825 where we have multiple constructors with the differing type(s).

**Comment by eslutz** [comment]
> This may be the same thing as the current failing test. But I was thinking we should have a unit test for this scenario - https://github.com/RockLib/RockLib.Configuration/blob/main/RockLib.Configuration.ObjectFactory/ConfigurationObjectFactory.cs#L825 where we have multiple constructors with the differing type(s).

Isn't the test I added already testing that?  It's comparing two constructors where the only different between them is parameter type. Or are you saying you want more variation in the types?

As for the [failing test](https://github.com/RockLib/RockLib.Configuration/blob/9f15407cad0569ac532f5fd8520099412b56bf99/Tests/RockLib.Configuration.ObjectFactory.Tests/ConfigurationObjectFactorySadPathTests.cs#L317-L335) that I missed, it's not going to throw an exception because I cleared up the ambiguity, so should the test be removed?  I can't think of a way to get it to pass.

**Comment by NeilSchiefer** [comment]
> > This may be the same thing as the current failing test. But I was thinking we should have a unit test for this scenario - https://github.com/RockLib/RockLib.Configuration/blob/main/RockLib.Configuration.ObjectFactory/ConfigurationObjectFactory.cs#L825 where we have multiple constructors with the differing type(s).
> 
> Isn't the test I added already testing that? It's comparing two constructors where the only different between them is parameter type. Or are you saying you want more variation in the types?
> 
> As for the [failing test](https://github.com/RockLib/RockLib.Configuration/blob/9f15407cad0569ac532f5fd8520099412b56bf99/Tests/RockLib.Configuration.ObjectFactory.Tests/ConfigurationObjectFactorySadPathTests.cs#L317-L335) that I missed, it's not going to throw an exception because I cleared up the ambiguity, so should the test be removed? I can't think of a way to get it to pass.

Ah right yeah that test can now be removed.  However, is there a test somewhere in the configrationObjectFactory*Tests that tests a similar two constructors differring by only type?  I just want to be sure our original use case would pass that IsInvokableWithDefaultParameters check.

**Comment by eslutz** [comment]
> > > This may be the same thing as the current failing test. But I was thinking we should have a unit test for this scenario - https://github.com/RockLib/RockLib.Configuration/blob/main/RockLib.Configuration.ObjectFactory/ConfigurationObjectFactory.cs#L825 where we have multiple constructors with the differing type(s).
> > 
> > 
> > Isn't the test I added already testing that? It's comparing two constructors where the only different between them is parameter type. Or are you saying you want more variation in the types?
> > As for the [failing test](https://github.com/RockLib/RockLib.Configuration/blob/9f15407cad0569ac532f5fd8520099412b56bf99/Tests/RockLib.Configuration.ObjectFactory.Tests/ConfigurationObjectFactorySadPathTests.cs#L317-L335) that I missed, it's not going to throw an exception because I cleared up the ambiguity, so should the test be removed? I can't think of a way to get it to pass.
> 
> Ah right yeah that test can now be removed. However, is there a test somewhere in the configrationObjectFactory*Tests that tests a similar two constructors differring by only type? I just want to be sure our original use case would pass that IsInvokableWithDefaultParameters check.

Yes, in the RockLib.Configuration.ObjectFactory.Tests project, I have this test,
```C#
[Theory]
[InlineData(2, 3, -1)]
[InlineData(3, 2, 1)]
[InlineData(2, 2, 0)]
public void CompareParameterTypes(int constructorOneIndex, int constructorTwoIndex, int expectedComparisonValue)
{
   var constructors = typeof(ConstructorsWithMatchingParamCount).GetConstructors();
   var members = new Dictionary<string, IConfigurationSection>
      {
          { "name", null! },
          { "url", null! }
      };
   var resolver = new Resolver(t => t, t => t == typeof(string) ? true : false);

   var orderInfo1 = _constructorOrderInfoType.New(constructors[constructorOneIndex], members, resolver);
   var orderInfo2 = _constructorOrderInfoType.New(constructors[constructorTwoIndex], members, resolver);

   Assert.Equal(expectedComparisonValue, orderInfo1.CompareTo(orderInfo2));
}
```
that compares these constructors.  I figured it made sense to test with the constructor parameters we were actually having trouble with lol.
```C#
public ConstructorsWithMatchingParamCount(string name, Uri url, string method = "POST", IReadOnlyDictionary<string, string>? defaultHeaders = null) { }
public ConstructorsWithMatchingParamCount(string name, string url, string method = "POST", IReadOnlyDictionary<string, string>? defaultHeaders = null) { }
```

**Comment by NeilSchiefer** [comment]
> > > > This may be the same thing as the current failing test. But I was thinking we should have a unit test for this scenario - https://github.com/RockLib/RockLib.Configuration/blob/main/RockLib.Configuration.ObjectFactory/ConfigurationObjectFactory.cs#L825 where we have multiple constructors with the differing type(s).
> > > 
> > > 
> > > Isn't the test I added already testing that? It's comparing two constructors where the only different between them is parameter type. Or are you saying you want more variation in the types?
> > > As for the [failing test](https://github.com/RockLib/RockLib.Configuration/blob/9f15407cad0569ac532f5fd8520099412b56bf99/Tests/RockLib.Configuration.ObjectFactory.Tests/ConfigurationObjectFactorySadPathTests.cs#L317-L335) that I missed, it's not going to throw an exception because I cleared up the ambiguity, so should the test be removed? I can't think of a way to get it to pass.
> > 
> > 
> > Ah right yeah that test can now be removed. However, is there a test somewhere in the configrationObjectFactory*Tests that tests a similar two constructors differring by only type? I just want to be sure our original use case would pass that IsInvokableWithDefaultParameters check.
> 
> Yes, in the RockLib.Configuration.ObjectFactory.Tests project, I have this test,
> 
> ```cs
> [Theory]
> [InlineData(2, 3, -1)]
> [InlineData(3, 2, 1)]
> [InlineData(2, 2, 0)]
> public void CompareParameterTypes(int constructorOneIndex, int constructorTwoIndex, int expectedComparisonValue)
> {
>    var constructors = typeof(ConstructorsWithMatchingParamCount).GetConstructors();
>    var members = new Dictionary<string, IConfigurationSection>
>       {
>           { "name", null! },
>           { "url", null! }
>       };
>    var resolver = new Resolver(t => t, t => t == typeof(string) ? true : false);
> 
>    var orderInfo1 = _constructorOrderInfoType.New(constructors[constructorOneIndex], members, resolver);
>    var orderInfo2 = _constructorOrderInfoType.New(constructors[constructorTwoIndex], members, resolver);
> 
>    Assert.Equal(expectedComparisonValue, orderInfo1.CompareTo(orderInfo2));
> }
> ```
> 
> that compares these constructors. I figured it made sense to test with the constructor parameters we were actually having trouble with lol.
> 
> ```cs
> public ConstructorsWithMatchingParamCount(string name, Uri url, string method = "POST", IReadOnlyDictionary<string, string>? defaultHeaders = null) { }
> public ConstructorsWithMatchingParamCount(string name, string url, string method = "POST", IReadOnlyDictionary<string, string>? defaultHeaders = null) { }
> ```

Well that only tests ConstructorOrderInfo.CompareTo.  What I mean is a test for ConfigurationObjectFactory.GetConstructor that tests with those same constructors.

**Comment by eslutz** [comment]
> > > > > This may be the same thing as the current failing test. But I was thinking we should have a unit test for this scenario - https://github.com/RockLib/RockLib.Configuration/blob/main/RockLib.Configuration.ObjectFactory/ConfigurationObjectFactory.cs#L825 where we have multiple constructors with the differing type(s).
> > > > 
> > > > 
> > > > Isn't the test I added already testing that? It's comparing two constructors where the only different between them is parameter type. Or are you saying you want more variation in the types?
> > > > As for the [failing test](https://github.com/RockLib/RockLib.Configuration/blob/9f15407cad0569ac532f5fd8520099412b56bf99/Tests/RockLib.Configuration.ObjectFactory.Tests/ConfigurationObjectFactorySadPathTests.cs#L317-L335) that I missed, it's not going to throw an exception because I cleared up the ambiguity, so should the test be removed? I can't think of a way to get it to pass.
> > > 
> > > 
> > > Ah right yeah that test can now be removed. However, is there a test somewhere in the configrationObjectFactory*Tests that tests a similar two constructors differring by only type? I just want to be sure our original use case would pass that IsInvokableWithDefaultParameters check.
> > 
> > 
> > Yes, in the RockLib.Configuration.ObjectFactory.Tests project, I have this test,
> > ```cs
> > [Theory]
> > [InlineData(2, 3, -1)]
> > [InlineData(3, 2, 1)]
> > [InlineData(2, 2, 0)]
> > public void CompareParameterTypes(int constructorOneIndex, int constructorTwoIndex, int expectedComparisonValue)
> > {
> >    var constructors = typeof(ConstructorsWithMatchingParamCount).GetConstructors();
> >    var members = new Dictionary<string, IConfigurationSection>
> >       {
> >           { "name", null! },
> >           { "url", null! }
> >       };
> >    var resolver = new Resolver(t => t, t => t == typeof(string) ? true : false);
> > 
> >    var orderInfo1 = _constructorOrderInfoType.New(constructors[constructorOneIndex], members, resolver);
> >    var orderInfo2 = _constructorOrderInfoType.New(constructors[constructorTwoIndex], members, resolver);
> > 
> >    Assert.Equal(expectedComparisonValue, orderInfo1.CompareTo(orderInfo2));
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > that compares these constructors. I figured it made sense to test with the constructor parameters we were actually having trouble with lol.
> > ```cs
> > public ConstructorsWithMatchingParamCount(string name, Uri url, string method = "POST", IReadOnlyDictionary<string, string>? defaultHeaders = null) { }
> > public ConstructorsWithMatchingParamCount(string name, string url, string method = "POST", IReadOnlyDictionary<string, string>? defaultHeaders = null) { }
> > ```
> 
> Well that only tests ConstructorOrderInfo.CompareTo. What I mean is a test for ConfigurationObjectFactory.GetConstructor that tests with those same constructors.

I finally get it.  It doesn't look like that test currently exists, so I'll work on getting it added.

**Comment by JasonBock** [comment]
What @NeilSchiefer said :). I feel like if we're trying to fix a bug in code, we should probably be adding a set of tests (at least 1). That's not a 100% truism, but that's typically how it goes. In fact, I usually write a test that fails to show the bug is happening, fix the bug, and then that test should pass as a result.

**Comment by JasonBock** [comment]
Also, just remember to update the version number to reflect a bug fix in your release.

File: RockLib.Configuration.ObjectFactory/ConstructorOrderInfo.cs [hunk 1]
```diff
@@ -56,6 +58,11 @@ public int CompareTo(ConstructorOrderInfo? other)
          if (TotalParameters < other.TotalParameters) return 1;
          if (MatchedNamedParameters > other.MatchedNamedParameters) return -1;
          if (MatchedNamedParameters < other.MatchedNamedParameters) return 1;
+         if (ParameterTypes is not null && other.ParameterTypes is not null)
+         {
+            if (ParameterTypes.Except(other.ParameterTypes).Any()) return -1;
+            if (other.ParameterTypes.Except(ParameterTypes).Any()) return 1;
```
Comment (JasonBock) [inline RockLib.Configuration.ObjectFactory/ConstructorOrderInfo.cs]: I'm wondering if there's a way to do this such that you don't have to do 4 passes over the enumerables. Offhand I don't have a suggestion, but....maybe just get the parameters as a list or array and then compare each one? Also, what happens if the parameter counts are different? Do we return 0 in that case, or 1/-1? And what's the right thing to do in that scenario?

**Review by JasonBock — CHANGES_REQUESTED** [review]
Just remember to update the version number to reflect that the release has a bug fix. I didn't see the

File: RockLib.Configuration.ObjectFactory/ConstructorOrderInfo.cs (hunk 1 above)
Comment (eslutz) [reply RockLib.Configuration.ObjectFactory/ConstructorOrderInfo.cs]: Updated it to a list.

File: Tests/RockLib.Configuration.ObjectFactory.Tests/ConfigurationObjectFactoryTests.cs [hunk 2]
```diff
@@ -14,7 +14,48 @@ namespace Tests
...
+             {
+                    { "MatchingConstructorsWithDifferenthParamType:name", "test" },
+                    { "MatchingConstructorsWithDifferenthParamType:url", "https://www.google.com" }
+             })
+             .Build();
+
+            var fooSection = config.GetSection("MatchingConstructorsWithDifferenthParamType");
+            var foo = fooSection.Create<MatchingConstructorsWithDifferenthParamType>()!;
+            Assert.Equal("test", foo.Name);
+            Assert.Equal(new Uri("https://www.google.com"), foo.Url);
```
Comment (NeilSchiefer) [inline Tests/RockLib.Configuration.ObjectFactory.Tests/ConfigurationObjectFactoryTests.cs:31]: Nice! It prefers the Uri version, exactly what we want.

File: RockLib.Configuration.ObjectFactory/CHANGELOG.md [hunk 3]
```diff
@@ -5,6 +5,12 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## Unreleased
+
+#### Changed
+
+- Updated `ToCompare()` method in `ConstructorOrderInfo` to also compare the parameter types during comparison.
```
Comment (NeilSchiefer) [inline RockLib.Configuration.ObjectFactory/CHANGELOG.md]: Oops typo in method name.  Should be CompareTo()

Comment (NeilSchiefer) [reply RockLib.Configuration.ObjectFactory/CHANGELOG.md]: Could be fixed in the release commit though.

**Review by NeilSchiefer — APPROVED** [review]

**Comment by NeilSchiefer** [comment]
Nice job!

File: RockLib.Configuration.ObjectFactory/CHANGELOG.md (hunk 3 above)
Comment (eslutz) [reply RockLib.Configuration.ObjectFactory/CHANGELOG.md]: Good point, but too late lol.

**Review by JasonBock — APPROVED** [review]

**Pull request merged**